### ISSUE TITLE
Foreign trait implementations cannot be instanced

### ISSUE THREAD

**Issue by threema-lenny** [mozilla/uniffi-rs#1845]
It seems that foreign trait implementations currently (0.25 and 0.25.1) don't work. Unfortunately, I cannot provide an easy reproducer that actually spits out a compile error but I think you'll be able to reproduce this easily, just bear with me:

Let's take a look at the [traits example](https://github.com/mozilla/uniffi-rs/tree/main/examples/traits). IIRC the current Kotlin binding is broken and it looks like the tests don't pick this up: While we can derive from `Button`, we cannot create an instance of it because `Button` requires a pointer which we don't have. Here is a snippet of the code that is being generated for the example.

```kotlin
class Button(
    pointer: Pointer
) : FFIObject(pointer), ButtonInterface {
    ...
}
```

We don't know `pointer` and there is no helper code generated to create `Button` either. So, [this](https://github.com/mozilla/uniffi-rs/blob/main/examples/traits/tests/bindings/test_traits.kts) won't work:

```kotlin
class KtButton : Button {
    override fun name() = "KtButton"
}
```

To me, this looks as if the code is being prepared so that a Rust implementation of the trait `Button` can be used with it but not a foreign implementation.

Edit: It is possible to work around this issue by using the _soft-deprecated_ callback interface instead.

Side question: Is there a sample repo (for each supported language) so that we can easily write reproducers that check for compile errors of the foreign code or possibly even run it?

**Comment by mhammond** [comment]
> It seems that foreign trait implementations currently (0.25 and 0.25.1) don't work. Unfortunately, I cannot provide an easy reproducer that actually spits out a compile error but I think you'll be able to reproduce this easily, just bear with me:

I'm afraid I can't. If I instrument this test I can see it working as expected.

> Here is a snippet of the code that is being generated for the example.

That's odd - I see instead:

```
public interface Button {
    fun `name`(): String
    companion object
}

open class ButtonImpl : FFIObject, Button {
    constructor(pointer: Pointer): super(pointer)
...
```

Maybe you are accidentally on an old version?

> Side question: Is there a sample repo (for each supported language) so that we can easily write reproducers that check for compile errors of the foreign code or possibly even run it?

That's what the `examples` and `fixtures` directories are for - you can run them directly from a uniffi-rs checkout. Patches to a fixture to show a problem/regression really is the best way to submit bugs.

**Comment by threema-lenny** [comment]
Huh... that's very odd, indeed. Okay, so I've made a repro here, referencing 0.25.1: https://github.com/threema-lenny/uniffi-fs-repros/tree/foreign-trait-issue

This is the relevant snippet of `build/kotlin/uniffi/lib/lib.kt` when I run `build.sh`:

```kotlin
class Button(
    pointer: Pointer
) : FFIObject(pointer), ButtonInterface {

    /**
     * Disconnect the object from the underlying Rust object.
     *
     * It can be called more than once, but once called, interacting with the object
     * causes an `IllegalStateException`.
     *
     * Clients **must** call this method once done with the object, or cause a memory leak.
     */
    override protected fun freeRustArcPtr() {
        rustCall() { status ->
            _UniFFILib.INSTANCE.uniffi_repros_fn_free_button(this.pointer, status)
        }
    }

    override fun `name`(): String =
        callWithPointer {
    rustCall() { _status ->
    _UniFFILib.INSTANCE.uniffi_repros_fn_method_button_name(it,

        _status)
}
        }.let {
            FfiConverterString.lift(it)
        }




    companion object

}
```

**Comment by mnorth719** [comment]
Wanted to chime in here as I was running into the same issue in Swift. 

It looks like this is no longer an issue on the latest commit on main ([main](https://github.com/mozilla/uniffi-rs/commit/eccc994f6b9903593dd075b899f489228ecd9c5e) at time of writing this). I am also able to reproduce the above issue on the latest published release `0.25.1`.

Good news is, looks like this will be addressed in the next published release. I'd recommend the callback workaround you mentioned or pointing at `main` in the meantime until this gets published.

**Comment by mhammond** [comment]
The next published release will hopefully be 0.25.2 - I'd love to know how exactly this was fixed to see if it might also be a candidate for that. Not sure I'll find the time to do that though (I'm struggling to find the time to put that release together as it is!)

**Comment by mhammond** [comment]
It looks like this and #1846 are actually dupes?

**Comment by bendk** [comment]
I wonder if I messed up the `0.25.1` release somehow, what happens if you run the code against the `v0.25.1` git branch?

**Comment by threema-lenny** [comment]
> I wonder if I messed up the `0.25.1` release somehow, what happens if you run the code against the `v0.25.1` git branch?

Ran it with `uniffi = { git = "https://github.com/mozilla/uniffi-rs.git", tag = "v0.25.1", ... }` but the result is the same.

Sanity-check: It does generate the code @mhammond saw when using `{ git = "https://github.com/mozilla/uniffi-rs.git", branch = "main", ...}`.

**Comment by mhammond** [comment]
It turns out the problem here is actually the changelog - allowing foreign implementations of traits was merged after 0.25, so it's expected to not work there. We'll update the changelog but you'll need to use main or wait for 0.26 for this capability.

If you do still have problems there, please make a new repro against one of our fixtures or examples and open a new issue.

Sorry for the confusion.

**Comment by ianthetechie** [comment]
Thanks for the update + xref @mhammond. I also came to the same conclusion and was just about to come over to post about this but missed the callout somehow 😂 

I'd love to be able to pin to a release so we can get the first version of our crate published (forgot that crates.io doesn't accept stuff that's pinned to a commit hash -_-), but it's not super urgent or anything. Master is working great for development.
